File: trOS_io/src/framebuffer.rs
use trOS_mbox::mbox::*;
use trOS_phys::mmio::write;

// @TODO: Remove this when SD card support lands.
static FONT: &'static [u8; 2080] = include_bytes!("font.psf");

/// Represent a PSF font. This struct is packed to allow the use of
/// `core::ptr::read`.
#[repr(C, packed)]
struct PSFFont {
    magic: u32,
    version: u32,
    headersize: u32,
    flags: u32,
    numglyph: u32,
    bytes_per_glyph: u32,
    height: u32,
    width: u32,
}

/// Represent the linear framebuffer of the RPI3. Note that this struct holds a
/// raw pointer, and unsafe code acts on this to write to the framebuffer. See
/// the `impl`'s of this function for more details.
pub struct FrameBufferStream {
    width: u32,
    height: u32,
    pitch: u32,
    x: u32,
    y: u32,
    ptr: *mut u8,
}

impl Default for FrameBufferStream {
    fn default() -> Self {
        FrameBufferStream{
            width: 1920,
            height: 1080,
            pitch: 0,
            x: 0,
            y: 0,
            ptr: 0xDEAD_BEEF as *mut u8,
        }
    }
}

impl FrameBufferStream {
    /// Instantiate a new `FrameBufferStream` with a given width and height.
    /// The `ptr` and `pitch` fields are instantiated with filler values.
    pub fn new(width: u32, height: u32) -> Self {
        FrameBufferStream{
            width: width,
            height: height,
            pitch: 0,
            x: 0,
            y: 0,
            ptr: 0xDEAD_BEEF as *mut u8,
        }
    }

    /// Initialize the linear framebuffer. Performs a mailbox call to set the
    /// resolution, then sets the values for `pitch` and `ptr` of the
    /// `FrameBufferStream` it was called on.
    pub fn init(&mut self, mbox: &mut MailBox) -> Result<(), &str>{
        // Clear the mailbox and setup the new call
        mbox.clear();
        mbox.mbox[0] = 35 * 4;
        mbox.mbox[1] = MBOX_REQUEST;
        mbox.mbox[2] = 0x48003; //set phy wh
        mbox.mbox[3] = 8;
        mbox.mbox[4] = 8;
        mbox.mbox[5] = self.width;
        mbox.mbox[6] = self.height;

        mbox.mbox[7] = 0x48004; //set virt wh
        mbox.mbox[8] = 8;
        mbox.mbox[9] = 8;
        mbox.mbox[10] = self.width;
        mbox.mbox[11] = self.height;

        mbox.mbox[12] = 0x48009; //set virt offset
        mbox.mbox[13] = 8;
        mbox.mbox[14] = 8;
        mbox.mbox[15] = 0;
        mbox.mbox[16] = 0;

        mbox.mbox[17] = 0x48005; //set depth
        mbox.mbox[18] = 4;
        mbox.mbox[19] = 4;
        mbox.mbox[20] = 32;

        mbox.mbox[21] = 0x48006; //set pixel order
        mbox.mbox[22] = 4;
        mbox.mbox[23] = 4;
        mbox.mbox[24] = 1;

        mbox.mbox[25] = 0x40001; //get framebuffer, gets alignment on request
        mbox.mbox[26] = 8;
        mbox.mbox[27] = 8;
        mbox.mbox[28] = 4096;
        mbox.mbox[29] = 0;

        mbox.mbox[30] = 0x40008; //get pitch
        mbox.mbox[31] = 4;
        mbox.mbox[32] = 4;
        mbox.mbox[33] = 0;

        mbox.mbox[34] = MBOX_TAG_LAST;

        if mbox.call(MBOX_CH_PROP) && mbox.mbox[20] == 32 && mbox.mbox[28] != 0 {
            mbox.mbox[28] &= 0x3FFF_FFFF;
            self.pitch = mbox.mbox[33];
            self.ptr = mbox.mbox[28] as *mut u8;
            Ok(())
        } else {
            Err("Failed to initialize framebuffer!")
        }
    }

    /// Write to the framebuffer at a given offset.
    fn write_at_offset(&mut self, offset: u32, d: u8) -> Result<(), &str> {
        if offset > 8_298_239 {
            return Err("WARNING: failed to write to offset, maximum offset is 8298239!");
        }
        unsafe {
            write((self.ptr.offset(offset as isize)) as *mut u32, d.into());
        }
        Ok(())
    }

    /// Clear the framebuffer to a given RGB color, in the most
    /// inneficient way possible.
    pub fn clear(&mut self, r: u8, g: u8, b: u8) {
        for y in 0..self.height {
            for x in 0..self.width * 2 {
                let offset = y * self.pitch + x * 3;
                self.write_at_offset(offset, r).unwrap();
                self.write_at_offset(offset + 1, g).unwrap();
                self.write_at_offset(offset + 2, b).unwrap();
            }
        }
    }

    pub fn write(&mut self, data: &str) {
        let font = unsafe { core::ptr::read(FONT.as_ptr() as *const PSFFont) };
        for c in data.chars() {
            let bytes_per_line = (font.width + 7) / 8;
            //let mut ptr = unsafe {(FONT.as_ptr()).offset(font.headersize as isize).offset(('A' as u32 * font.bytes_per_glyph) as isize)};
            let mut ptr = unsafe {(FONT.as_ptr()).offset(font.headersize as isize).offset((c as u32 * font.bytes_per_glyph) as isize)};
            let mut offset = (self.y * font.height * self.pitch) + (self.x * (font.width + 1) * 4);
            if c == '\r' {
                self.x = 0;
            } else if c == '\n' {
                self.x = 0;
                self.y += 1;
            } else {
                for _ in 0..font.height {
                    let mut line = offset;
                    let mut mask = 1 << (font.width - 1);
                    for _ in 0..font.width {
                        let color = unsafe { if *ptr & mask == 0 { 0 } else { 255 }};
                        self.write_at_offset(line, color).unwrap();
                        self.write_at_offset(line + 1, color).unwrap();
                        self.write_at_offset(line + 2, color).unwrap();
                        mask >>= 1;
                        line += 4;
                    }
                    ptr = unsafe {ptr.offset(bytes_per_line as isize)};
                    offset += self.pitch;
                }
                self.x += 1;
            }
        }
    }
}

File: src/vga/framebuffer.zig
const std = @import("std");
const io = @import("../io.zig");
const types = @import("../types.zig");

const gpio = io.gpio;
const mmio = io.mmio;
const mbox = io.mbox;
const uart = io.uart;

const Register = types.regs.Register;
const NoError = types.errorTypes.NoError;

const fontEmbed = @embedFile("font.psf");

/// Respresent a PSF font.
const PSFFont = packed struct {
    magic: u32,
    version: u32,
    headersize: u32,
    flags: u32,
    numglyph: u32,
    bytesPerGlyph: u32,
    height: u32,
    width: u32,
};

/// FrameBufferStream is a struct that handles framebuffer writes.
/// Note that it also handles parsing PSF fonts and doing glyph
/// lookups when performing a write.
pub const FrameBuffer = struct {
    const Self = @This();

    var width: u32 = undefined;
    var height: u32 = undefined;
    var pitch: u32 = undefined;
    // @NOTE: This should work well for on-the-fly font changes.
    // For example, we could start with a non-unicode font, then
    // swap to one if the need arises and continue printing seemlessly.
    var font: *const PSFFont = @ptrCast(*const PSFFont, &fontEmbed);
    var ptr: [*]volatile u8 = undefined;

    var column: u32 = 0;
    var row: u32 = 0;

    var initState = false;


    fn init() ?void {
        mbox.mbox[0] = 35*4;
        mbox.mbox[1] = mbox.MBOX_REQUEST;
        mbox.mbox[2] = 0x48003;  //set phy wh
        mbox.mbox[3] = 8;
        mbox.mbox[4] = 8;
        mbox.mbox[5] = 1920;
        mbox.mbox[6] = 1080;

        mbox.mbox[7] = 0x48004;  //set virt wh
        mbox.mbox[8] = 8;
        mbox.mbox[9] = 8;
        mbox.mbox[10] = 1920;
        mbox.mbox[11] = 1080;

        mbox.mbox[12] = 0x48009; //set virt offset
        mbox.mbox[13] = 8;
        mbox.mbox[14] = 8;
        mbox.mbox[15] = 0;
        mbox.mbox[16] = 0;

        mbox.mbox[17] = 0x48005; //set depth
        mbox.mbox[18] = 4;
        mbox.mbox[19] = 4;
        mbox.mbox[20] = 32;

        mbox.mbox[21] = 0x48006; //set pixel order
        mbox.mbox[22] = 4;
        mbox.mbox[23] = 4;
        mbox.mbox[24] = 1;

        mbox.mbox[25] = 0x40001; //get framebuffer, gets alignment on request
        mbox.mbox[26] = 8;
        mbox.mbox[27] = 8;
        mbox.mbox[28] = 4096;
        mbox.mbox[29] = 0;

        mbox.mbox[30] = 0x40008; //get pitch
        mbox.mbox[31] = 4;
        mbox.mbox[32] = 4;
        mbox.mbox[33] = 0;

        mbox.mbox[34] = mbox.MBOX_TAG_LAST;

        if ((mbox.mboxCall(mbox.MBOX_CH_PROP) != null) and mbox.mbox[20] == 32 and mbox.mbox[28] != 0) {
            mbox.mbox[28] &= 0x3FFFFFFF;
            width = mbox.mbox[5];
            height = mbox.mbox[6];
            pitch = mbox.mbox[33];
            ptr = @intToPtr([*]volatile u8, mbox.mbox[28]);
        } else {
            return null;
        }
    }

    fn put(c: u8) void {
        const bytesPerLine = (font.width + 7) / 8;
        var offset = (row * font.height * pitch) + (column * (font.width + 1) * 4);
        var idx = usize(0);
        switch(c) {
            '\r' => {
                writeBytes("\n");
                writeBytes("READY:> ");
            },
            '\n' => {
                column = 0;
                row += 1;
            },
            // Backspace
            8 => {
                if (column > 8)
                    column -= 1;
                offset = (row * font.height * pitch) + (column * (font.width + 1) * 4);
                var y: usize = 0;
                while (y < font.height) : (y += 1) {
                    var line = offset;
                    var x: usize = 0;
                    while (x < font.width) : (x += 1) {
                        ptr[line] = 0;
                        ptr[line + 1] = 0;
                        ptr[line + 2] = 0;
                        line += 4;
                    }
                    offset += pitch;
                }
            },
            else => {
                if (c < font.numglyph) {
                    idx += (font.headersize + (c * font.bytesPerGlyph));
                } else {
                    idx += (font.headersize + (0 * font.bytesPerGlyph));
                }
                var y: usize = 0;
                while (y < font.height) : (y += 1) {
                    var line = offset;
                    var mask = u32(1) << @truncate(u5, (font.width - 1));
                    var x: usize = 0;
                    while (x < font.width) : (x += 1) {
                        var color: u8 = undefined;
                        if ((fontEmbed[idx]) & mask == 0) {
                            color = 0;
                        } else {
                            color = 255;
                        }
                        ptr[line] = color;
                        ptr[line + 1] = color;
                        ptr[line + 2] = color;
                        mask >>= 1;
                        line += 4;
                    }
                    idx += bytesPerLine;
                    offset += pitch;
                }
                column += 1;
            },
        }
    }

    pub fn writeBytes(data: []const u8) void {
        if (!initState)
            init().?;
        for (data) |c| {
            put(c);
        }
    }
};


/// `put` is a public wrapper around the `put` function contained within
/// `FrameBufferStream`.
pub fn put(c: u8) void {
    FrameBuffer.put(c);
}

/// `writeHandler` handles write requests for the framebuffer from `write`.
fn writeHandler(context: void, data: []const u8) NoError!void {
    FrameBuffer.writeBytes(data);
}

/// `write` manages all writes for the framebuffer. It takes formatted arguments, in the
/// same manner that `std.debug.warn()` does. It then passes them to `writeHandler`
/// for writing out.
pub fn write(comptime data: []const u8, args: ...) void {
    std.fmt.format({}, NoError, writeHandler, data, args) catch |e| switch (e) {};
}

// @PENDING-FIX: Test fails with: "TODO buf_read_value_bytes packed struct"
//test "PSFFont read test" {
//    const font = @ptrCast(*const PSFFont, &fontEmbed);
//    uart.write("{}", font);
//    std.debug.assert(font.magic == 2253043058);
//    std.debug.assert(font.version == 0);
//    std.debug.assert(font.headersize == 32);
//    std.debug.assert(font.flags == 786432);
//    std.debug.assert(font.numglyph == 128);
//    std.debug.assert(font.bytesPerGlyph == 16);
//    std.debug.assert(font.height == 16);
//    std.debug.assert(font.width == 8);
//}
